[FIX] payment_*: create redirect urls from http request

ap-wtioit opened the pull request:
to prevent redirecting the user to web.base.url when his request did
come from a different domain we use the domain in given in the http
request to create the redirect urls for the payment providers

Description of the issue/feature this PR addresses:
When a user makes a payment over an URL that is not web.base.url he gets redirected to web.base.url instead of the original url (e.g. Shopping over www2.example.com, paying over paypal, the user then gets redirected to www.example.com instead of www2.example.com after the payment)

Current behavior before PR:
After a finished payment a user always gets redirected to web.base.url

Desired behavior after PR is merged:
After a finished payment a user is redirected to the original domain

fixes #11638

Info @wt-io-it, @dbo-odoo 

this PR replaces https://github.com/odoo/odoo/pull/33439

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

ap-wtioit commented:
force pushed /after rebasing against current odoo 10.0 branch (and to get a new testrun from runbot (could not reproduce test failure locally))

wtaferner commented:
ping @dbo-odoo

JKE-be commented:
@kebeclibre + f727cd inspiration

dbo-odoo commented on addons/payment/models/payment_acquirer.py:
```diff
@@ -367,6 +368,9 @@ def button_immediate_install(self):
             }
 
+    def _get_base_url(self):
+        return request and request.httprequest.url_root or self.env['ir.config_parameter'].get_param('web.base.url')
```
probaly needs a `sudo` for the `get_param` call

dbo-odoo commented:
Hi @ap-wtioit,

This is more or less what was done in 937b5c076e7175bec664e for version 12.0; since version 12.0 supports multiple websites, the problem was more important there.

Your patch looks good, could you retarget version 11.0, if you still want it? Besides a little correction I included as a comment.

TY

ap-wtioit commented:
Created pull request against 11.0 #43361, also renamed the method to have the same name as in 12.0 937b5c076e7175bec664ed0cf4b77505e342f1e2 for hopefully easier forward porting.

Also added the suggested sudo() for version 11.0 (i think it was not needed in 10.0)

dbo-odoo commented:
I'm not even sure the sudo is really required, since I think we're already in a sudoed env from some place up in the callstack. Better safe than sorry though, and it's really weird to have a `get_param` without sudo starting from 11.0 (before that it relied on groups set on the config param itself, so perhaps it was indeed not needed in 10.0)

ap-wtioit commented:
merged in 11.0 see #43361
